### ISSUE TITLE
Parallelize downloading multiple packages sources

### ISSUE THREAD

**Issue by kangalioo** [Jguer/yay#915]
Would it be possible to add support for downloading and building in parallel? I had in mind that when installing multiple packages, yay could already start building the first package while the second package is downloading. Then the second package is built while the third is downloaded and so on.

I think that would be a nice feature to have because downloading and building can be parallelized so well. It would make updating the system with `yay -Syu` up to twice as fast.

**Comment by Jguer** [comment]
Parallel building is not a great advantage because then you have competition for cycles (and packages with dependencies will have to be treated). Concurrent download is something that has been thought about and experimented with (it messes a bit with the outputs though).

**Comment by kangalioo** [comment]
True, parallel building wouldn't be beneficial, but that's not what I meant. I meant having the downloading process running in parallel to the building process. Those process need completely different parts of the system, so parallelizing those two would provide a big speed improvement.

**Comment by Morganamilo** [comment]
I can see it working but I can't see any way to implement it while also keeping any sort of readable output.

**Comment by kangalioo** [comment]
It could be a command-line option that's turned off by default, so yay has readable output by default. When enabled, yay could display a status message in between the compilation output giving information about the downloads, maybe something like this:
```
...
[ 41%] Building CXX object ...
[yay] Download of ___.tar.gz finished, starting download ___.tar.gz...
[ 43%] Building CXX object ...
[ 45%] Building CXX object ...
...
```

**Comment by BrendanBall** [comment]
Seen as parallelizing builds isn't an option, trying to download and build in parallel seems like overkill. I would go with just parallelizing multiple downloads at the same time.

**Comment by dsdante** [comment]
@Jguer why not pass -j to make and let it do parallelization?

@BrendanBall what's the point in dividing the bandwidth among packages?

**Comment by Jguer** [comment]
@dsdante That's the job of `/etc/makepkg.conf` not yay, you already have the parallelization of a single build and that is effective (everything C++ for example ...). Parallelization of multiple builds would be like trying to shove 20 clowns by a door at a single time, probably slower than having them line up. 

Dividing the bandwidth among packages could be beneficial for multiple AUR builds as the sources are different and there could be slow servers in some of them. 
On repo packages where all the packages have the same source your connection will typically max out before you're able to max out a single repository, so less effective.

**Comment by dsdante** [comment]
@Jguer thanks for the makepkg tip!

I think the easiest way to optimize it is to make a pipeline:
A queue of clowns -> downloading -> building -> installing (but why clowns, anyway?)

The downloader can let the next package use some of the bandwith if it's not 100% used yet and if it's from a different server. Not sure it would worth your time though. (I don't speak Go, so I'm pass, sorry.)

Another issue is output. We'd need to provide info from multiple downloads, building (which is parallelized itself) and installing. I can't think of a nice solution without ncurses.

**Comment by SicariusNoctis** [comment]
> I can see it working but I can't see any way to implement it while also keeping any sort of readable output.

One idea is split windows:

```
[Build output]
==> Starting pkgver()...
==> Starting build()...
Submodule 'submodules/HaRe' (https://github.com/alanz/HaRe.git) registered for path 'submodules/HaRe'
Submodule 'submodules/brittany' (https://github.com/lspitzner/brittany.git) registered for path 'submodules/brittany'
Submodule 'submodules/cabal-helper' (https://github.com/alanz/cabal-helper.git) registered for path 'submodules/cabal-helper'
Submodule 'submodules/ghc-mod' (https://github.com/alanz/ghc-mod.git) registered for path 'submodules/ghc-mod'
Cloning into '/tmp/yaytmp-1000/haskell-ide-engine-git/src/haskell-ide-engine/submodules/HaRe'...
Cloning into '/tmp/yaytmp-1000/haskell-ide-engine-git/src/haskell-ide-engine/submodules/brittany'...
Cloning into '/tmp/yaytmp-1000/haskell-ide-engine-git/src/haskell-ide-engine/submodules/cabal-helper'...

---

[Download bars]
[ ===        ]   30% downloaded alacritty-git
[ ====       ]   40% downloaded aurman
```

**Comment by kangalioo** [comment]
I'm not sure if we're all talking about the same thing here...

@SicariusNoctis From your suggestions it seems like you're thinking of doing multiple processes of the same type at the same time. For example that both `alacritty-git` and `aurman` are downloaded simultaneously, while `cargo-git` and `haskell-ide-engine-git` are built simultaneously (stealing each other CPU cycles).

What I really had in mind though was a pipeline, as @dsdante said. There would be one downloading unit and one compiling unit that work in parallel. But each of these units only ever processes one item at a time (similar to how modern CPU pipelines work).

**Comment by ForeverZer0** [comment]
This seems like it would add a lot of complexity and open up an avenue for lots of bugs, and all for virtually no significant benefit. 

Off the top of my head I can think of multiple scenarios and edge cases that would cause this to break. Even without any errors, it would simply be messy and add complexity. Popping up multiple windows to show progress, or mixing output from a download in with the build output is just ugly no matter how you do it.

15 years ago when we didn't have the bandwidth we do now, and building took much longer on ancient hardware this may have been more important to consider, but do we really need this now?

**Comment by phiresky** [comment]
I agree that it would be pretty complex, but I don't think the "bandwidth" argument is valid. 

Right now I'm sitting here waiting for yay to update and it's downloading `intel-mkl`, which is 3GB in size which even with 10MByte per second takes 5 minutes to download. If the downloads were parallelized, it could pretty much could have installed all the other updates during this time.

**Comment by Hubro** [comment]
I definitely think each package should be downloaded and built in separate processes.

When I'm updating 10-20 packages, the CPU or network bandwidth is *barely* being hit all the way through (unless an update requires rebuilding Chromium or something). It's generally just a long queue of "downloading... unpacking... making package... tidying install... cleaning up..." 20 times in a row and it takes a good 20 minutes. The update would have taken perhaps 2 minutes if the download/build/make-package part of each package was done in a separate process.

The main process could simply display a progress bar for each package. This would be an amazing improvement.

**Comment by kangalio** [comment]
I think there's a lot of misunderstanding here about my original proposal. First of all, what was bugging me and led me to make this issue is that yay is only ever using CPU _or_ bandwidth at a time. When yay is downloading a package, it's wasting time by letting the CPU idle and vice versa.

I did not propose to run multiple builds in parallel. Neither to run two downloads in parallel. Those things are not necessarily beneficial.

The goal was to have a pipeline with two steps: a downloading step and a build step. When a package has finished downloading, it'll be moved into the build step, and the next package can already start downloading. This ensures that CPU and bandwidth are utilized continuously.

As far as I can tell, the only contra argument that's applicable to the _actual_ feature I'm proposing is the terminal output issue. The issue is that when a download runs alongside a build, their log messages will mix. YodaEmbedding experimented with three possible ways to handle this issue in [their comment](https://github.com/Jguer/yay/issues/915#issuecomment-495026441). In my opinion, those implementations are a bit over-engineered, and a [simpler solution](https://github.com/Jguer/yay/issues/915#issuecomment-477239055) would be more appropriate in a command-line tool like yay.

Additionally, I think it would be good to have the pipeline feature opt-in. That way, users are not exposed to the potentially confusing behavior if they don't explicitly request it.

Are there any complaints about this way of handling terminal output?

**Comment by EmeraldSnorlax** [comment]
Sorry for the necro; now that pacman 6.0 has released and pacman supports parallel downloads, ensuring that your connection is saturated at all times (particularly useful for many small packages), now seems like a better time than ever for having parallel downloads and building in yay.

I have a particularly slow internet connection; while building a package that takes a while, it'd save *so much time* to download the sources for packages in the background while a compilation for a different package is happening.

```
DL PKG A ---> compile ---> compress, tidy, etc ----> done
              DL PKG B -----> DL'd ---(wait)-------> compile, etc
                              DL PKG C --------------------------->
```

i.e. have a download queue and a build queue: only one download or build happens at a time, but the two queues happen in parallel. When a package is downloaded, it gets moved to the build queue and the next item in the download queue starts downloading.

From what I've gathered above, the main issue seems to be "there's no clean way of displaying output" -- honestly, my internet connection is so slow that I don't care if it's ugly, as long as I can download a bunch of other packages while waiting for some large C program to build. Maybe it'd make sense to not have it on by default if UX is a concern, until a sensible way of showing progress is thought up with. (Honestly: why even show progress for package downloads that are happening in the background, unless the build queue is empty and it's waiting for a package?)

**Comment by Jguer** [comment]
Implemented in #1573, currently it's a bit ugly visually but it works :sweat_smile:

**Comment by manorit2001** [comment]
> Implemented in #1573, currently it's a bit ugly visually but it works sweat_smile

https://github.com/gosuri/uiprogress something like this for better outputs might be good I think

**Comment by jargon4220** [comment]
How to enable this? yay is downloading repo for 30 minutes and all build and arch package downloads and installations have been suspended in the meantime.
